ircd/privilege.c

```diff
@@ -156,40 +326,118 @@ privilegeset_unref(struct PrivilegeSet *set)
 	{
 		rb_dlinkDelete(&set->node, &privilegeset_list);
 
-		rb_free(set->name);
-		rb_free(set->privs);
-		rb_free(set);
+		privilegeset_free(set);
+	}
+}
+
+const struct PrivilegeSet **
+privilegeset_diff(const struct PrivilegeSet *old, const struct PrivilegeSet *new)
+{
+	static const char *no_privs[] = { NULL };
+	static const struct PrivilegeSet empty = { .size = 0, .privs = no_privs };
+	static struct PrivilegeSet *set_unchanged = NULL,
+	                           *set_added = NULL,
+							   *set_removed = NULL;
+	static const struct PrivilegeSet *result_sets[3];
+	static size_t n_privs = 0;
+	size_t new_size = n_privs ? n_privs : 32;
+	size_t i = 0, j = 0;
+
+	if (result_sets[0] == NULL)
+	{
+		result_sets[0] = set_unchanged = privilegeset_new_orphan("<unchanged>");
+		result_sets[1] = set_added = privilegeset_new_orphan("<added>");
+		result_sets[2] = set_removed = privilegeset_new_orphan("<removed>");
+	}
+
+	if (old == NULL)
+		old = &empty;
+	if (new == NULL)
+		new = &empty;
+
+	while (new_size < MAX(old->size, new->size) + 1)
+		new_size *= 2;
+
+	if (new_size > n_privs)
+	{
+		set_unchanged->privs = rb_realloc(set_unchanged->privs, sizeof *set_unchanged->privs * new_size);
+		set_added->privs = rb_realloc(set_added->privs, sizeof *set_added->privs * new_size);
+		set_removed->privs = rb_realloc(set_removed->privs, sizeof *set_removed->privs * new_size);
 	}
+
+	const char **res_unchanged = set_unchanged->privs;
+	const char **res_added = set_added->privs;
+	const char **res_removed = set_removed->privs;
+
+	while (i < old->size || j < new->size)
+	{
+		const char *oldpriv = NULL, *newpriv = NULL;
+		int ord = 0;
+		if (i < old->size)
```

Comment:
We shouldn't need these conditionals. The priv lists are NULL terminated, so i and j can just get stuck pointing to the NULL at the end.